libqtile/backend/x11/core.py

```diff
@@ -600,25 +601,60 @@ def handle_ClientMessage(self, event) -> None:  # noqa: N802
             except IndexError:
                 logger.debug("Invalid desktop index: %s" % index)
 
+    def fake_KeyPress(self, keycode, needs_release=True) -> None:
+        xtest = self.conn.conn(xcffib.xtest.key)
+        # The key needs to be released as it was previously pressed
+        if needs_release:
+            xtest.FakeInput(3, keycode, xcffib.xproto.Time.CurrentTime, 0, 0, 0, 0)
+            self.flush()
+        # Press the key
+        xtest.FakeInput(2, keycode, xcffib.xproto.Time.CurrentTime, 0, 0, 0, 0)
+        self.flush()
+        # Release the key
+        xtest.FakeInput(3, keycode, xcffib.xproto.Time.CurrentTime, 0, 0, 0, 0)
+        self.flush()
+
     def handle_KeyPress(self, event) -> None:  # noqa: N802
         assert self.qtile is not None
 
         keysym = self.conn.code_to_syms[event.detail][0]
-        self.qtile.process_key_event(keysym, event.state & self._valid_mask)
+        key, handled = self.qtile.process_key_event(keysym, event.state & self._valid_mask)
+
+        # As we're grabbing async we can't just replay it, so...
+        # We need to forward the event to the focused window
+        if not handled and key:
+            # We need to ungrab the key as otherwise we get an event loop
+            self.ungrab_key(key)
+            # Modifier is already pressed, just repeat the keycode
+            self.fake_KeyPress(event.detail)
+            # Grab the key again
+            self.grab_key(key)
```

Comment:
@m-col could it be the case that we're re-grabbing the key too early? Or does the `flush()` call already take care of it? I'm not sure if this is right what I'm doing.